diagnostics: Current file diagnostics view

dinocosta opened the pull request:
These changes introduce a new command to the Diagnostics panel, `diagnostics: deploy current file`, which allows the user to view the diagnostics only for the currently opened file. 

Here's a screen recording showing these changes in action 🔽 

[diagnostics: deploy current file](https://github.com/user-attachments/assets/b0e87eea-3b3a-4888-95f8-9e21aff8ea97)

Closes #4739 

Release Notes:

- Added new `diagnostics: deploy current file` command to view diagnostics for the currently open file

dinocosta committed:
feat(diagnostics): add buffer diagnostics editor

Introduce a new diagnostics view, `BufferDiagnosticsEditor`, which
allows users to view the diagnostics for the currently open file. A new
action has also been introduced so users can deploy this view from the
command palette, `diagnostics: deploy current file`.

This view is very similar in behaviour to `ProjectDiagnosticsEditor`,
but it was worth noting that, while `ProjectDiagnosticsEditor` reads and
writes to the global `IncludeWarnings` setting,
`BufferDiagnosticsEditor` only reads its value when first being created,
otherwise it keeps its own `include_warnings` state so as to not affect
the `ProjectDiagnosticsEditor` or other `BufferDiagnosticsEditor`
entities.

A new method was introduced to make it easier to retrieve the diagnostic
summary for a specific path rather for all paths,
`project::lsp_store::LspStore.diagnostic_summary_for_path`, seeing as
`BufferDiagnosticsEditor` only needs to display the summary for a single
path.

dinocosta committed:
test: add buffer diagnostics editor tests

Update the `diagnostics::diagnostics_tests` module with three new tests
for the `BufferDiagnosticsEditor` view:

- `test_buffer_diagnostics` - Asserts that the `BufferDiagnosticsEditor`
displays diagnostics only for the provided project path, ignoring other
paths with diagnostics.
- `test_buffer_diagnostics_without_warnings` - Asserts that if
`include_warnings` is `false`, the `BufferDiagnosticsEditor` displays
only errors in the editor, without showing warnings.
- `test_buffer_diagnostics_multiple_servers` - Asserts that if the
project path has multiple diagnostics for different language servers,
all diagnostics are shown and included in the summary.

Additionally, the
`project::lsp_store::LspStore.diagnostic_summary_for_path` method has
been updated to return the total count of warnings and errors across all
language servers for the provided path, rather than just the first
language server. This change aligns it with the behavior when
diagnostics are displayed, as all language servers are considered.

This commit also updates the `BufferDiagnosticsEditor::deploy` function
in order to fetch the buffer directly from the active editor and provide
it to `BufferDiagnosticsEditor::new`. This avoids having to spawn the
background task that would open the buffer and update the
`BufferDiagnosticsEditor.buffer` field.

ConradIrwin commented:
@dinocosta I think that all errors from all language servers should be considered (so the summary should be the total), but looks like you implemented that already?

I think you can go one step further and pass the buffer directly from the editor instead of re-openinig it in the constructor.

Finally, though I didn't figure out why, `g space` (or alt-enter) to open the current excerpt in the main file don't seem to work in the buffer diagnostics view.

I talked with @SomeoneToIgnore and we can safely ignore the cargo stuff (he'd like to remove it completely, but we don't need to do that as part of this PR).

dinocosta commented:
@ConradIrwin Thank you for reviewing again and providing valuable feedback! 🙇

> @dinocosta I think that all errors from all language servers should be considered (so the summary should be the total), but looks like you implemented that already?

Yes, I forgot to mention this. While adding tests for `BufferDiagnosticsEditor` in [this commit](https://github.com/zed-industries/zed/pull/34430/commits/ba7f155f4a3e6d281e8bbbf58722719e21053479), I updated the implementation of `project::lsp_store::LspStore::diagnostic_summary_for_path` to return the total number of errors and warnings. Since the diagnostics editor displays all diagnostics from all language servers, it made sense for the summary to consider all language servers as well.

> I think you can go one step further and pass the buffer directly from the editor instead of re-openinig it in the constructor.

I was trying to figure this out because I suspected I should be able to call the `as_singleton` method on the editor's multibuffer to get the correct entity for use in `BufferDiagnosticsEditor`. I believe the changes in https://github.com/zed-industries/zed/pull/34430/commits/bd626555f84284f2de02da830d98b9c9faeaca69 have addressed this issue. We no longer need to spawn a task to open the buffer!

> Finally, though I didn't figure out why, `g space` (or alt-enter) to open the current excerpt in the main file don't seem to work in the buffer diagnostics view.

Thank you for bringing this to my attention! I discovered that some methods for the `workspace::item::Item` trait were not implemented for `BufferDiagnosticsEditor`. I have now implemented the same methods that `ProjectDiagnosticsEditor` uses for the `Item` trait in https://github.com/zed-industries/zed/pull/34430/commits/d9c1ea1f4480cd58f82a833adcdab8d0d4c0f9e3. As a result, `g space` is now functioning as intended, including the ability to split the buffer diagnostics.

> I talked with @SomeoneToIgnore and we can safely ignore the cargo stuff (he'd like to remove it completely, but we don't need to do that as part of this PR).

Awesome! Those have been completely removed from `BufferDiagnosticsEditor` in https://github.com/zed-industries/zed/pull/34430/commits/f0f7b53cc82e81be82ad18f7c121c48ccafe62c1 🧹

---

I believe this is now ready for another review, I'll also go ahead and update the Pull Request's title and description, as it no longer matches the latest changes 🙂

dinocosta committed:
refactor: introduce diagnostics toolbar editor trait

This commit refactors the interaction between the diagnostics editors
(`BufferDiagnosticsEditor` and `ProjectDiagnosticsEditor`) so that,
instead of passing an optional enum to the `ToolbarControls` struct, as
well as the `DiagnosticBlock`, we pass a reference to a type that
implements a new trait, `DiagnosticsToolbarEditor`, which defines the
methods that the toolbar controls needs, namely:

- `include_warnings`
- `toggle_warnings`
- `has_stale_excerpts`
- `is_updating`
- `stop_updating`
- `refresh_diagnostics`
- `get_diagnostics_for_buffer`

This makes it so that both `ToolbarControls` and `DiagnosticBlock`
implementations don't need to be aware of the internal details of each
of the editors, as well as avoiding duplicating the
`DiagnosticsEditorHandle` enum.

ConradIrwin commented:
thanks @SomeoneToIgnore! And sorry @dinocosta, this conflicts. Are you able to rebase?

dinocosta commented:
> > we can safely ignore the cargo stuff (he'd like to remove it completely, but we don't need to do that as part of this PR).
> 
> Removed in
> 
> #36795

Thank you @SomeoneToIgnore ! 🙇 

> thanks @SomeoneToIgnore! And sorry @dinocosta, this conflicts. Are you able to rebase?

@ConradIrwin No worries! I've rebased these changes with `main` and ended up pushing another commit – https://github.com/zed-industries/zed/pull/34430/commits/67b3f80003a8e745b6508aa6920c7906ad85fb20 – refactoring the `DiagnosticsEditorHandle` enums into a `DiagnosticsToolbarEditor` trait so that `ToolbarControls` does not need to know that there's two different diagnostics editors. This could probably be renamed and moved since it's actually being used by both `diagnostics::toolbar_controls` and `diagnostics::diagnostic_render` so feedback is appreciated! 🙏

dinocosta committed:
chore: use only filename in tab content

Co-authored-by: Conrad Irwin <conrad.irwin@gmail.com>
